### PR TITLE
LP-1877 Remove three days old referral issue resolved

### PR DESCRIPTION
**Pull request by muak50**
### Description

[LP-1877](https://philanthropyu.atlassian.net/browse/LP-1877)

- Filter query issue resolved for 3 days old referrals.
- Duplicate emails issue resolved.

### CODE REVIEW

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py [hunk 1]
```diff
@@ -5,35 +5,40 @@
...
+from openedx.features.smart_referral.tasks import task_send_referral_follow_up_emails
 
 CONTACT_EMAIL = 'contact_email'
 DAYS_TO_SEND_FOLLOW_UP_EMAIL = 3
 
 
 class Command(BaseCommand):
     help = """
-    Send follow up referral email to all those contacts
+    Send follow-up referral email to all those contacts
```
Comment (harris-irfan) [inline common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:16]: Please add single indent in line 16.

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py [hunk 2]
```diff
@@ -5,35 +5,40 @@
...
     """
 
     def handle(self, *args, **options):
         emails_list = []
-        three_days_old_date = timezone.now().date() - timedelta(days=DAYS_TO_SEND_FOLLOW_UP_EMAIL)
+        referral_threshold_date = timezone.now() - timedelta(days=DAYS_TO_SEND_FOLLOW_UP_EMAIL)
 
-        three_days_old_referral = SmartReferral.objects.filter(
-            created__date__lte=three_days_old_date,
+        three_days_old_referral_emails = SmartReferral.objects.filter(
```
Comment (harris-irfan) [inline common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:23]: we can update the variable name to referral_threshold_date_emails to make it generic so that if the number of days change in future, we only have to change the value in the constant and not the variable name. Also, it will make it consistent with the referral_threshold_date variable.

File: common/djangoapps/philu_commands/management/commands/tests/test_send_referral_follow_up_emails.py [hunk 3]
```diff
@@ -18,8 +18,8 @@ class SendReferralFollowUpEmails(TestCase):
         Tests for `send_referral_follow_up_email` command.
     """
 
-    @mock.patch('philu_commands.management.commands.send_referral_follow_up_email.task_send_referral_follow_up_email')
-    def test_send_referral_follow_up_command(self, mock_task_send_referral_follow_up_email):
+    @mock.patch('philu_commands.management.commands.send_referral_follow_up_email.task_send_referral_follow_up_emails')
+    def test_send_referral_follow_up_emails_command(self, mock_task_send_referral_follow_up_email):
         """
         This test will return a list of email address. The email address will follow certain criteria. We are getting
         emails from 'SmartReferral' tables so only those entries will be picked that are at-least three days old. If an
```
Comment (harris-irfan) [inline common/djangoapps/philu_commands/management/commands/tests/test_send_referral_follow_up_emails.py:25]: I think the docstring is not appropriate. The test does not return anything. Maybe it can be rephrased to:

"This test verifies that a call is made to the function 'task_send_referral_follow_up_email' with a list of email address. The email address will follow certain criteria. We are getting emails from 'SmartReferral' table such that the referrals are at least 3 days old. If an email address is registered on our platform (PhilU) it will be ignored. Lastly, if the 'is_referral_step_complete' flag is 'True', that email will also be ignored. "

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py [hunk 4]
```diff
@@ -5,35 +5,40 @@
...
-        ).values(
-            CONTACT_EMAIL
+        ).values_list(
+            CONTACT_EMAIL,
+            flat=True
         )
-        for referral in three_days_old_referral:
+
+        # Removing duplicate email entries.
+        three_days_old_referral_emails = list(set(three_days_old_referral_emails))
```
Comment (zahid-arbisoft) [inline common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:32]: If possible lets use `.distinct()` in the query rather than using set to remove duplicates

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py [hunk 5]
```diff
@@ -5,35 +5,40 @@
 from django.utils import timezone
 
 from openedx.features.smart_referral.models import SmartReferral
-from openedx.features.smart_referral.tasks import task_send_referral_follow_up_email
+from openedx.features.smart_referral.tasks import task_send_referral_follow_up_emails
 
 CONTACT_EMAIL = 'contact_email'
```
Comment (harris-irfan) [inline common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:10]: We don't need this constant anymore since 'contact_email' is now used only once in this file, and the string itself is self-explanatory.

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 1 above)
Comment (harris-irfan) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:16]: Plus, the docstring is not explanatory enough. We need to better define 'all those contacts'

Maybe we can rephrase it to:

'Send follow-up referral email to all pending contacts'

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 4 above)
Comment (harris-irfan) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:32]: Upon using `.distinct()`, we can also remove the comment as it would be obvious that we are removing duplicates with the use of distinct.

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 1 above)
Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:16]: done

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 2 above)
Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:23]: This variable scope is limited to the command. It wouldn't affect our system. So if in future the number of days would be changed we will change the variable without any complication.  It won't have any ripple effects on our system. So I guess 'three_days_old_referral_emails' name is more understandable and explain it's purpose.

File: common/djangoapps/philu_commands/management/commands/tests/test_send_referral_follow_up_emails.py (hunk 3 above)
Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/tests/test_send_referral_follow_up_emails.py:25]: done

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 5 above)
Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:10]: Still we are avoiding string, I guess there's no harm in doing this.

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 4 above)
Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:32]: Thanks, done

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 5 above)
Comment (zeshanarifios) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:10]: What I remember in clean code sessions we read that use magic constants for things which are not clear .. like 3 we don't know what would be 3 ..so it's better to use constant for 3 .. 
but contact_email is self explanatory string.. so i think we can avoid this constant..

Comment (zeshanarifios) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:10]: Secondly constant `CONTACT_EMAIL` and string `'contact_email'` are almost same.. So does this really benefit in terms of reusability even if it would be used on multiple places?

Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:10]: @zeshanarifios Please review, [https://docs.google.com/document/d/11jqz9wpX4twsrFy5k6jlAazO7oTeQPqr415ZojmBAFM/edit#heading=h.fj7jz7a37jgn](Code review checklist)

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py (hunk 2 above)
Comment (zeshanarifios) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:24]: I also agree with @harris-irfan .. It would be great if we can make this variable name generic here..

Comment (muak50) [reply common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py]: done

File: common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py [hunk 6]
```diff
@@ -5,35 +5,38 @@
...
 DAYS_TO_SEND_FOLLOW_UP_EMAIL = 3
 
 
 class Command(BaseCommand):
     help = """
-    Send follow up referral email to all those contacts
+    Send follow-up referral email to all those contacts who are at-least 3 days old, contacts which aren't registered
+    on our platform and we hadn't send them follow-up email before.
```
Comment (harris-irfan) [inline common/djangoapps/philu_commands/management/commands/send_referral_follow_up_email.py:17]: The sentence has some errors. It can be rephrased to:

"Send follow-up referral email to all those contacts which are at least 3 days old, are not registered on our platform, and have not received follow-up email."
